### PR TITLE
Handle schemes other than the file scheme

### PR DESCRIPTION
**Pull request by rwols**
This is a draft where I'm updating the code to handle schemes other than `file`.

Points of concern:

- [x] goto def should handle arbitrary URIs
- [x] goto references should handle arbitrary URIs
- [x] restore output panel functionality of goto references
- [x] what about workspace symbols?
- ~~if a view is already open with the given URI, switch to that view
      instead of opening a new view.~~ in another PR
- [x] arbitrary URI schemes are handled through AbstractPlugin.on_open_uri_async
- [x] related infos for diagnostics should work
- ~~SessionBuffer (or maybe DocumentSyncListener) should be aware of what
      its URI is. It currently only knows its filename. Most likely, this will
      result in us also being able to handle buffer-only views.~~ in another PR
- [x] jump back history

### CODE REVIEW

**Comment by rwols** [comment]
So, textDocument/references understands arbitrary URIs now as well:

https://user-images.githubusercontent.com/2431823/111921072-dea2c680-8a92-11eb-8871-ab58fd6babca.mov

And if an URI is not a file:// URI it goes through the AbstractPlugin extension point.

**Comment by rwols** [comment]
Example on how to handle custom uris: https://github.com/sublimelsp/LSP-Deno/blob/1b5cda88f0c182424c0c1b14b36f6cafff8ec344/plugin.py#L14-L26

File: plugin/core/views.py [hunk 1]
```diff
@@ -158,7 +155,24 @@ def region_to_range(view: sublime.View, region: sublime.Region) -> Range:
...
+def get_uri_and_range_from_location(location: Union[Location, LocationLink]) -> Tuple[DocumentUri, RangeLsp]:
+    if "targetUri" in location:
+        location = cast(LocationLink, location)
+        uri = location["targetUri"]
+        r = location["targetSelectionRange"]
+    else:
+        location = cast(Location, location)
+        uri = location["uri"]
+        r = location["range"]
+    return uri, r
```
[the remark this answers is not in the document]
Comment (rwols) [reply plugin/core/views.py:171]: I don't understand this... there are no errors?

[the remark this answers is not in the document]
Comment (rchl) [reply plugin/core/views.py:171]: Hmm, maybe it was some version of Pyright that had an issue with that...

**Comment by rwols** [comment]
> Because on some places we use weak refs, on some places we don't and I cannot figure out when is expected to use weak refs, or when there is no need for them.

I'll write a wiki page about it if I can find the time. But basically, just never hold on to a session object. A `sublime_plugin.TextCommand` is instantiated once per view, and then kept alive for the duration of the view. Therefore, if you stash something in such a command via `self.foo = bar`, then the command will keep a reference to `bar` as long as the view lives.

You can test this by printing something in `__init__` and `__del__` of a text command.

**Comment by predragnikolic** [comment]
I find the reference panel more useful than showing references in the quick panel.

Screenshot 2021-05-08 at 15 37 52

Pros of the reference panel:
- Shows the symbol we are finding reference for
- The count of references
- A preview of each line
- If you want to go to a reference you click it, or press a keybinding

Cons of the reference panel:
- could not handle custom URIs


Screenshot 2021-05-08 at 15 36 43

Pros of the quick panel:
- Can handle custom URIs.

Cons of the quick panel:
- Doesn't shows the symbol we are finding reference for
- Doesn't shows the count of references
- While it doesn't shows a preview of each line in quick panel, if we press UP/DOWN arrows it will take us to that line, where we could see he line, unless the reference is at the top of the file... than the quick panel will cover it, so we won't be able to see it

For me the pros and cons of the reference panel are better that 
the pros and cons of the quick panel.

Suggestion:
Show the references in the panel, and don't rely on ST result_file_regex, result_line_regex that are required for the `next_result` `prev_result` to work. 
Instead add custom commands for navigation `lsp_next_result`, `lsp_prev_result` that would allow handling of custom URIs.

-----
Also I noticed that the "show_references_in_quick_panel" was removed 9adb815 ,
but there is still one place in the code where that setting is just read:
https://github.com/sublimelsp/LSP/blob/st4000-exploration/plugin/core/types.py#L171
https://github.com/sublimelsp/LSP/blob/st4000-exploration/plugin/core/types.py#L203

**Comment by rwols** [comment]
> Instead add custom commands for navigation lsp_next_result, lsp_prev_result that would allow handling of custom URIs.

This would still not allow clicking on the URI. As for the disadvantages of the quick panel:

> Doesn't shows the symbol we are finding reference for

I guess it's possible to make use of the targetSelectionRange for LocationLinks and range for Locations.

> Doesn't shows the count of references

Is that really important information?

> While it doesn't shows a preview of each line in quick panel, if we press UP/DOWN arrows it will take us to that line, where we could see he line, unless the reference is at the top of the file... than the quick panel will cover it, so we won't be able to see it

The built-in Goto References also has this problem. I've been running with this branch for about a month now and it's honestly not that big of a problem.

Is it okay to at least roll with this for a while and see how we can improve it further? Given that panels don't understand URIs, I don't think panels have any future. That's really the bigger issue.

**Comment by rchl** [comment]
> Given that panels don't understand URIs

I currently don't have opinion on the panel vs. quick panel but I don't think that that statement is true. A custom panel together with a custom mouse binding that triggers a custom command that parses the clicked text should be possible.

**Comment by rwols** [comment]
It’s too much work/maintenance for me to maintain two very different presentations of the same information. So let’s go with the panel solution then and see if we can make it work.

**Comment by rchl** [comment]
The quick panel has the major advantage (IMO) that it's more keyboard-friendly. I can trigger `Find references` from the command palette, then filter results and go to a specific one I want. With the panel I have to go through all results with F4 which a) can open many files unnecessarily b) can take longer to find the right one if there are many results.

**Review by predragnikolic — DISMISSED** [review]
Thanks for keeping the reference panel! :)

**Pull request merged**